File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.js
function solution(edges) {
  const indegree = {};
  const outdegree = {};
  const nodes = new Set();

  // 간선 정보 저장
  for (const [u, v] of edges) {
    nodes.add(u);
    nodes.add(v);

    outdegree[u] = (outdegree[u] || 0) + 1;
    indegree[v] = (indegree[v] || 0) + 1;

    // 없는 키 초기화
    indegree[u] = indegree[u] || 0;
    outdegree[v] = outdegree[v] || 0;
  }

  let stickCount = 0;
  let eightCount = 0;
  let genNode = 0;

  // 모든 노드 순회
  for (const node of nodes) {
    const inDeg = indegree[node];
    const outDeg = outdegree[node];

    if (inDeg === 0 && outDeg >= 2) {
      genNode = node; // 생성 정점
    } else if (outDeg === 0 && inDeg >= 1) {
      stickCount++;
    } else if (inDeg >= 2 && outDeg >= 2) {
      eightCount++;
    }
  }

  const dounutCount = outdegree[genNode] - (stickCount + eightCount);

  return [genNode, dounutCount, stickCount, eightCount];
}

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.py
from collections import defaultdict
def solution(edges):
    graph = defaultdict(list)
    indegree = defaultdict(int)
    outdegree = defaultdict(int)
    
    for u,v in edges:
        graph[u].append(v)
        outdegree[u] += 1
        indegree[v] += 1
        
        if not outdegree[v]:
            outdegree[v] = 0
        if not indegree[u]:
            indegree[u] = 0
 
    n = len(indegree.keys())
    
    # 중심, 막대기, 8자 탐색
    center = -1
    stick_count = 0
    eight_count = 0
    doughnut_count = 0
    for i in range(1, n + 1):
        if outdegree[i] >= 2 and indegree[i] == 0:
            center = i
        elif outdegree[i] == 0 and indegree[i] >= 1: 
            stick_count += 1
        elif outdegree[i] >= 2 and indegree[i] >= 2:
            eight_count += 1
    doughnut_count = outdegree[center] - stick_count - eight_count
    answer = [center, doughnut_count, stick_count, eight_count]
    print(answer)
    return answer